### PR TITLE
NOISSUE - Updating Things and Channels Service

### LINKED ISSUE #15: Update JS SDK for Things

### PR DESCRIPTION
**Pull request by wambui-pixel**
### What does this do? 

Updating Things .ts

### Which issue(s) does this PR fix/relate to? NOISSUE

Resolves #15 and #17

### List any changes that modify/break current functionality

Migrating things.js to things.ts

### Have you included tests for your changes?

NO

### Did you document any new/modified functionality?

NO

### CODE REVIEW

File: src/things.ts [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+    }
+
+    try {
+      const response = await fetch(
+        new URL(
+          `${this.thingsEndpoint}/${thing.id}/channels?${new URLSearchParams(
+            stringParams
+          ).toString()}`,
+          this.thingsUrl
+        ).toString(),
```
Comment (Musilah) [inline src/things.ts:235]:
```suggestion
        new URL(
          `channels${chan.id}/things?${new URLSearchParams(
            stringParams
          ).toString()}`,
          this.thingsUrl
        ).toString(),
```

File: src/things.ts [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const channelsData = await response.json()
+      return channelsData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async ThingsGetAll (
```
Comment (Musilah) [inline src/things.ts:249]: Whats the difference between this and `GetAll`?

File: src/things.ts [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+    // Disables thing.
+    /**
+     * @method Disable - Deletes a thing when provided with a valid token and thing ID.
+     * @param {string} thing_id - Thing ID.
+     * @param {string} token - User token.
+     * @returns {Object} - Thing object with statys disabled.
+     */
+
+    const options: RequestInit = {
+      method: 'post',
```
Comment (Musilah) [inline src/things.ts:304]:
```suggestion
      method: 'POST',
```

File: src/things.ts [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+          `${this.thingsEndpoint}/${thing.id}/disable`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const thingData = await response.json()
```
Comment (Musilah) [inline src/things.ts:323]:
```suggestion
      const thingData: Thing = await response.json()
```

File: src/things.ts [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async Disable (thing: Thing, token: string): Promise<Thing> {
+    // Disables thing.
+    /**
+     * @method Disable - Deletes a thing when provided with a valid token and thing ID.
+     * @param {string} thing_id - Thing ID.
```
Comment (Musilah) [inline src/things.ts:298]: update the docstrings

File: src/things.ts [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const thingData = await response.json()
+      return thingData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async Update (thing: Thing, token: string): Promise<Thing> {
```
Comment (Musilah) [inline src/things.ts:330]:
```suggestion
  public async UpdateThing (thing: Thing, token: string): Promise<Thing> {
```

File: src/things.ts [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+     * }
+     */
+
+    const options: RequestInit = {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify(thing)
```
Comment (Musilah) [inline src/things.ts:411]:
```suggestion
      body: JSON.stringify( {secret: thing.credentials?.secret} )
```

File: src/things.ts [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const thingData = await response.json()
+      return thingData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async IdentifyThing (thingKey: string): Promise<Thing> {
```
Comment (Musilah) [inline src/things.ts:523]: This should return a thingId in an `identifyThingResp` struct

File: src/things.ts [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+      method: 'POST',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Thing ${thingKey}`
+      }
+    }
+
+    try {
+      const response = await fetch(
+        new URL(this.thingsEndpoint, this.thingsUrl).toString(),
```
Comment (Musilah) [inline src/things.ts:534]:
```suggestion
        new URL('/identify', this.thingsUrl).toString(),
```

File: src/things.ts [hunk 10]
```diff
@@ -0,0 +1,839 @@
...
+        new URL(
+          `${this.thingsEndpoint}/${thingID}/permissions`,
+          this.thingsUrl
+        ).toString(), options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const thingData: Permissions = await response.json()
```
Comment (Musilah) [inline src/things.ts:584]: match the promise to what you're returning here

File: src/things.ts [hunk 11]
```diff
@@ -0,0 +1,839 @@
...
+      return thingData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async Enable (thing: Thing, token: string): Promise<Thing> {
+    // Enables thing.
+    /**
+     * @method Enable - Deletes a thing when provided with a valid token and thing ID.
```
Comment (Musilah) [inline src/things.ts:594]: Mismatch in the docstrings

File: src/things.ts [hunk 12]
```diff
@@ -0,0 +1,839 @@
...
+        throw this.thingError.HandleError(errorRes.error, response.status)
+      }
+      const thingData = await response.json()
+      return thingData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async Things (
```
Comment (Musilah) [inline src/things.ts:628]: This is a repeat of the `ThingsGetAll` and `GetAll` functions

Comment (Musilah) [reply src/things.ts:628]: Choose one and delete the rest

File: src/things.ts [hunk 13]
```diff
@@ -0,0 +1,839 @@
...
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async ListThingUsers (
+    thingId: string,
+    queryParams: QueryParams,
+    token: string
+  ): Promise<Users> {
```
Comment (Musilah) [inline src/things.ts:684]:
```suggestion
  ): Promise<UsersPage> {
```

File: src/things.ts [hunk 14]
```diff
@@ -0,0 +1,839 @@
...
+    // Retrieves list of users connected to specified thing with pagination metadata.
+    /**
+     * @method ListThingUsers - Retrieves list of users connected to specified thing
+     * with pagination metadata.
+     * @param {string}
```
Comment (Musilah) [inline src/things.ts:689]: update these docstrings too

**Review by Musilah — CHANGES_REQUESTED** [review]

File: src/channels.ts [hunk 15]
```diff
@@ -0,0 +1,757 @@
...
+      return channelData
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async ChannelPermissions (
+    channel: Channel,
+    token: string
+  ): Promise<any> {
```
Comment (Musilah) [inline src/channels.ts:360]:
```suggestion
  ): Promise<: Permissions> {
```

File: src/channels.ts [hunk 16]
```diff
@@ -0,0 +1,757 @@
...
+  public async Delete (channel: Channel, token: string): Promise<string> {
+    // Deletes channel with specified id.
+    /**
+     * @method Disable - Deletes channel with specified id.
+     * @param {Object} channel - Channel object with new information.
+     * @param {String} token - An access token that is valid.
+     * @returns {Object} - Channel Object.
+     */
+    const options: RequestInit = {
+      method: 'post',
```
Comment (Musilah) [inline src/channels.ts:475]:
```suggestion
      method: 'DELETE',
```

File: src/channels.ts [hunk 17]
```diff
@@ -0,0 +1,757 @@
...
+      throw error
+    }
+  }
+
+  public async ListChannelUsersGroups (
+    channelId: string,
+    queryParams: QueryParams,
+    token: string
+  ): Promise<GroupsPage> {
+    // Lists users in a channel.
```
Comment (Musilah) [inline src/channels.ts:504]: Update the docstrings.

File: src/channels.ts [hunk 18]
```diff
@@ -0,0 +1,757 @@
...
+      throw error
+    }
+  }
+
+  public async Connect (
+    thingsID: string,
+    channelsID: string,
+    token: string
+  ): Promise<Response> {
+    // Connects multiple things to multiple channels.
```
Comment (Musilah) [inline src/channels.ts:584]: Match the comments to the function.

File: src/channels.ts [hunk 19]
```diff
@@ -0,0 +1,757 @@
...
+     */
+    const Connection = {
+      thing_id: thingsID,
+      channel_id: channelsID
+    }
+    // const payload = {
+    //   subjects: thingsIDs,
+    //   objects: channelsIDs,
+    //   actions
+    // }
```
Comment (Musilah) [inline src/channels.ts:603]:
```suggestion
```

File: src/channels.ts [hunk 20]
```diff
@@ -0,0 +1,757 @@
...
+      throw error
+    }
+  }
+
+  public async Disconnect (
+    thingID: string,
+    channelID: string,
+    token: string
+  ): Promise<Response> {
+    // Connects thing to channel.
```
Comment (Musilah) [inline src/channels.ts:633]: Match the docstrings to the function here as well.

File: src/channels.ts [hunk 21]
```diff
@@ -0,0 +1,757 @@
...
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async ListChannelUsers (
+    channelId: string,
+    queryParams: QueryParams,
+    token: string
+  ): Promise<Users> {
```
Comment (Musilah) [inline src/channels.ts:723]:
```suggestion
  ): Promise<UsersPage> {
```

File: src/channels.ts [hunk 22]
```diff
@@ -0,0 +1,757 @@
...
+     * @method Connect - Connects thing to channel when provided with a valid token,
+     * channel id and a thing id. The thing must have an action that it can perform over
+     * the channel.
+     * @param {string} thing_id - Thing ID.
+     * @param {string} channel_id - Channel ID.
+     * @param {list} actions - Action for example: ["m_read", "m_write"].
+     * @param {string} token - User token.
+     *
+     */
+    const payload = { subject: thingId, object: channelId }
```
Comment (Musilah) [inline src/channels.ts:554]:
```suggestion
```

File: src/channels.ts [hunk 23]
```diff
@@ -0,0 +1,757 @@
...
+    const options = {
+      method: 'post',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify(payload)
```
Comment (Musilah) [inline src/channels.ts:561]:
```suggestion
```

File: src/channels.ts [hunk 24]
```diff
@@ -0,0 +1,757 @@
...
+      const shareResponse: Response = { status: response.status, message: 'Thing Connected Successfully' }
+      return shareResponse
+    } catch (error) {
+      throw error
+    }
+  }
+
+  public async Connect (
+    thingsID: string,
+    channelsID: string,
```
Comment (Musilah) [inline src/channels.ts:581]:
```suggestion
    channelID: string,
```

File: src/channels.ts [hunk 25]
```diff
@@ -0,0 +1,757 @@
...
+      thing_id: thingID,
+      channel_id: channelID
+    }
+    const options: RequestInit = {
+      method: 'post',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify(connIDs)
```
Comment (Musilah) [inline src/channels.ts:654]:
```suggestion
      body: JSON.stringify( { thing_id: thingID, channel_id: channelID } )
```

File: src/channels.ts [hunk 26]
```diff
@@ -0,0 +1,757 @@
...
+     * @param {string} thing_id - Thing ID.
+     * @param {string} channel_id - Channel ID.
+     * @param {list} actions - Action for example: ["m_read", "m_write"].
+     * @param {string} token - User token.
+     *
+     */
+    const connIDs = {
+      thing_id: thingID,
+      channel_id: channelID
+    }
```
Comment (Musilah) [inline src/channels.ts:647]:
```suggestion
```

File: src/channels.ts [hunk 27]
```diff
@@ -0,0 +1,757 @@
...
+     * @param {list} channel_ids - Array of channel IDs.
+     * @param {list} actions - Array of actions for example: ["m_read", "m_write"].
+     * @param {string} token - User token.
+     * @returns {Object} - Policy object.
+     *
+     */
+    const payload = {
+      subjects: thingId,
+      objects: channelId
+    }
```
Comment (Musilah) [inline src/channels.ts:693]:
```suggestion
```

File: src/channels.ts [hunk 28]
```diff
@@ -0,0 +1,757 @@
...
+    const options = {
+      method: 'post',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify(payload)
```
Comment (Musilah) [inline src/channels.ts:700]:
```suggestion
```

File: examples/things.ts [hunk 29]
```diff
@@ -0,0 +1,188 @@
+// Import the SDK class from the mainflux-sdk package
+import SDK from '../src/sdk'
+
+const defaultUrl = 'http://localhost'
+
+const mySdk = new SDK({
+  thingsUrl: defaultUrl + ':9000'
+})
+
+// // //Things.js examples.
```
Comment (Musilah) [inline examples/things.ts:10]:
```suggestion
// Things.js examples.
```

File: examples/things.ts [hunk 30]
```diff
@@ -0,0 +1,188 @@
...
+  })
+  .catch((error) => {
+    console.log(error)
+  })
+
+mySdk.things
+  .ShareThing(
+    '<thingId>',
+
+    'administrator',
```
Comment (Musilah) [inline examples/things.ts:133]:
```suggestion
    '<relation>',
```

**Review by Musilah — CHANGES_REQUESTED** [review]

File: src/users.ts [hunk 31]
```diff
@@ -163,6 +160,7 @@ export default class Users {
         ).toString(),
         options
       )
+      console.log(response)
```
Comment (ianmuchyri) [inline src/users.ts:163]: remove log

File: src/sdk.ts [hunk 32]
```diff
@@ -1,25 +1,55 @@
 import Users from './users'
+import Domains from './domains'
 import Things from './things'
+// import Groups from "./groups";
```
Comment (ianmuchyri) [inline src/sdk.ts:4]: remove all the commented out code.

File: src/defs.ts [hunk 33]
```diff
@@ -57,9 +55,14 @@ export interface GroupsPage {
 export interface Channel {
   id?: string
   name?: string
+  parent_id?: string
+  credentials?: {
```
Comment (ianmuchyri) [inline src/defs.ts:59]: a channel does not have credentials. 
Channel struct looks like this:
```
type Channel struct {
	ID          string     `json:"id,omitempty"`
	DomainID    string     `json:"domain_id,omitempty"`
	ParentID    string     `json:"parent_id,omitempty"`
	Name        string     `json:"name,omitempty"`
	Description string     `json:"description,omitempty"`
	Metadata    Metadata   `json:"metadata,omitempty"`
	Level       int        `json:"level,omitempty"`
	Path        string     `json:"path,omitempty"`
	Children    []*Channel `json:"children,omitempty"`
	CreatedAt   time.Time  `json:"created_at,omitempty"`
	UpdatedAt   time.Time  `json:"updated_at,omitempty"`
	Status      string     `json:"status,omitempty"`
	Permissions []string   `json:"permissions,omitempty"`
}

```

Comment (ianmuchyri) [reply src/defs.ts:59]: refer to the go sdk for these items

File: src/defs.ts [hunk 34]
```diff
@@ -34,6 +29,9 @@ export interface Thing {
   owner?: string
   metadata?: Record<string, any>
   status?: string
+  createdAt?: string
```
Comment (ianmuchyri) [inline src/defs.ts:32]: Lets use snake case for these

File: src/defs.ts [hunk 35]
```diff
@@ -18,14 +18,9 @@ export interface UsersPage {
   page: PageRes
 }
 
-export interface PageRes {
```
Comment (ianmuchyri) [inline src/defs.ts:21]: why are you deleting this interface?

File: src/channels.ts [hunk 36]
```diff
@@ -0,0 +1,695 @@
...
+     *  "metadata": {
+     *       "domain": "example.com"
+     *  },
+     * "status": "enabled",
+     * "owner_id": "bb7edb32-2eac-4aad-aebe-ed96fe073879"
+     * }
+     *
+     */
+    const options: RequestInit = {
+      method: 'post',
```
Comment (ianmuchyri) [inline src/channels.ts:65]: Keep the method names in capital for all
```suggestion
      method: 'POST,
```

File: src/channels.ts [hunk 37]
```diff
@@ -0,0 +1,695 @@
...
+     *
+     */
+    const payload = { subject: thingId, object: channelId }
+    const options = {
+      method: 'post',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify(payload)
```
Comment (ianmuchyri) [inline src/channels.ts:508]:
```suggestion
      body: JSON.stringify({ subject: thingId, object: channelId })
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: examples/channels.ts [hunk 38]
```diff
@@ -0,0 +1,202 @@
...
+  })
+  .catch((error) => {
+    console.log(error)
+  })
+
+mySdk.channels
+  .AddUserToChannel(
+    '<channelId>',
+    [
+      '<thingId>', '<thingId>'
```
Comment (ianmuchyri) [inline examples/channels.ts:176]:
```suggestion
      '<userId>', '<userId>'
```

File: examples/channels.ts [hunk 39]
```diff
@@ -0,0 +1,202 @@
...
+  })
+  .catch((error) => {
+    console.log(error)
+  })
+
+mySdk.channels
+  .RemoveUserFromChannel(
+    '<channelId>',
+    [
+      '<thingId>', '<thingId>'
```
Comment (ianmuchyri) [inline examples/channels.ts:192]:
```suggestion
      '<userId>', '<userId>'
```

File: src/channels.ts [hunk 40]
```diff
@@ -0,0 +1,684 @@
...
+     * @param {string} token - User token.
+     * @returns Response - 'Thing Connected Successfully'.
+     */
+    const options = {
+      method: 'POST',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify({ subject: thingId, object: channelId })
```
Comment (ianmuchyri) [inline src/channels.ts:547]:
```suggestion
      body: JSON.stringify({ thing_id: thingId, channel_id: channelId })
```

File: src/channels.ts [hunk 41]
```diff
@@ -0,0 +1,684 @@
...
+     * @param {string} token - User token.
+     *
+     */
+    const options: RequestInit = {
+      method: 'POST',
+      headers: {
+        'Content-Type': this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify({ subject: thingId, object: channelId })
```
Comment (ianmuchyri) [inline src/channels.ts:586]:
```suggestion
      body: JSON.stringify({ thing_id: thingId, channel_id: channelId })
```

File: src/channels.ts [hunk 42]
```diff
@@ -0,0 +1,684 @@
...
+          `channels/${channelId}/groups?${new URLSearchParams(stringParams).toString()}`,
+          this.usersUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const usersData: GroupsPage = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:480]:
```suggestion
      const groupsPage: GroupsPage = await response.json()
```

File: src/channels.ts [hunk 43]
```diff
@@ -0,0 +1,684 @@
...
+          `${this.channelsEndpoint}/${channelId}/permissions`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: Permissions = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:345]:
```suggestion
      const permissions: Permissions = await response.json()
```

File: src/channels.ts [hunk 44]
```diff
@@ -0,0 +1,684 @@
...
+          `${this.channelsEndpoint}/${channel.id}/enable`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: Channel = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:307]:
```suggestion
      const enabledChannel: Channel = await response.json()
```

File: src/channels.ts [hunk 45]
```diff
@@ -0,0 +1,684 @@
...
+          `${this.channelsEndpoint}/${channel.id}/disable`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: Channel = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:273]:
```suggestion
      const disabledChannel: Channel = await response.json()
```

File: src/channels.ts [hunk 46]
```diff
@@ -0,0 +1,684 @@
...
+          `channels/${channel.id}`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: Channel = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:239]:
```suggestion
      const updatedChannel: Channel = await response.json()
```

File: src/channels.ts [hunk 47]
```diff
@@ -0,0 +1,684 @@
...
+          ).toString()}`,
+          this.thingsUrl
+        ).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: ChannelsPage = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:158]:
```suggestion
      const channels: ChannelsPage = await response.json()
```

File: src/channels.ts [hunk 48]
```diff
@@ -0,0 +1,684 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.channelsEndpoint}/${channelId}`, this.thingsUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.channelError.HandleError(errorRes.error, response.status)
+      }
+      const channelData: Channel = await response.json()
```
Comment (ianmuchyri) [inline src/channels.ts:114]:
```suggestion
      const channel: Channel = await response.json()
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

**Review by ianmuchyri — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**